content/posts/2020/02/thai-in-2562/index.eng.md

```diff
@@ -0,0 +1,93 @@
+
+In Indochina Peninsula, Buddhism takes the same place as Christianity in Europe. So, instead of the [Gregorian calendar], which uses the birth of Jesus Christ as an epoch, countries in Indochina Peninsula had used the [Buddhist calendar] which takes Buddha’s attainment of parinirvana (nirvana after death) as an epoch. Although countries like Cambodia and Laos eventually took on the Gregorian calendar, Thailand still uses the [Thai solar calendar], which is a solar modification of the traditional Buddhist lunar calendar. Buddha’s parinirvana was in 543 BC, and so the year 2019 A.D. becomes 2562 under the Thai solar calendar.
+
+As such, the world uses various calendars depending on the cultural region. So when displaying user interface, time should be displayed in the appropriate date format for each region. In fact, `DateTimeOffset.ToString()` method has an overload that also receives [`IFormatProvider`][IFormatProvider] objects as parameters for this purpose. [`CultureInfo`][CultureInfo] is the most common class to implement the `IFormatProvider` interface. As the name suggests, `CultureInfo` is the same concept that the Unix family calls locale. As shown below, the result of `DateTimeOffset.ToString()` method depends on which locale you set up as the parameter.
+
+```csharp
+> using System.Globalization;
+> var now = DateTimeOffset.Now;
+> now.ToString("yyyy-MM-ddTHH:mm:ss.ffffffZ", new CultureInfo("ko-KR"))
+"2020-02-13T17:37:16.436163Z"
+> now.ToString("yyyy-MM-ddTHH:mm:ss.ffffffZ", new CultureInfo("th-TH"))
+"2563-02-13T17:37:16.436163Z"
+```
+
+However, if you omit the parameter without setting any locale, the method will follow the locale of the environment in which the code is executed. The code below is the result of setting up the operating system region in Korea.
+
+```csharp
+> now.ToString("yyyy-MM-ddTHH:mm:ss.ffffffZ")
+"2020-02-13T17:37:16.436163Z"
+```
+
+[According to the docs, the overload with the omitted parameter][2] follows [`CultureInfo.CurrentCulture`][CultureInfo.CurrentCulture]. As you can infer from its name, the `CultureInfo.CurrentCulture` property points to the locale of the execution environment. Therefore, you must explicitly specify [`CultureInfo.InvariantCulture`][CultureInfo.InvariantCulture] if you want a deterministic action at all times regardless of the locale of your execution environment.
+
+Even though the method may be non-deterministic, the API designed to follow the locale of the execution environment is probably intended because such formatting operations are usually used for user interface, and coding can naturally look appropriate to the cultural community without much concern about internationalization. But the reason why we used this method was not for user interface, but for the cryptographic hash input that had to be deterministic— this turned out to be a mistake.
```

Comment:
```suggestion
Even though the method may be non-deterministic, the API designed to follow the locale of the execution environment is probably intended because such formatting operations are usually used for a user interface, and coding can naturally look appropriate to the cultural community without much concern about internationalization. But the reason why we used this method was not for a user interface, but for the cryptographic hash input that had to be deterministic— this turned out to be a mistake.
```